### ISSUE TITLE
paths from inside irods_environment.json are relative to cwd

### ISSUE THREAD

**Issue by mcast** [irods/irods#4852]
## Bug Report

It is either a very minor / inconvenience bug report, or more of a feature request.

### iRODS Version, OS and Version

Seen with clients 4.1.12 and 4.2.7 on Ubuntu Bionic, with packages from https://irods.org/download/ repurposed across Ubuntu releases (details available if relevant)

### What did you try to do?

Copy an `irods_environment-dev.json`, file
* from "internal Linux" where I have username `mca` home `/nfs/users/nfs_m/mca`
* to a virtual machine where I have username `ubuntu` home `/home/ubuntu`

I have wrapper scripts named `i-$foo` which put the `-$foo` into `IRODS_ENVIRONMENT_FILE=/home/ubuntu/.irods/irods_environment-$foo.json` before running the rest of the command. Therefore I have multiple `.irodsA-$foo` files.

I imagine that as iRODS developers you have a similar need for rapidly and conveniently speaking to different iRODS zones or federations of zones?  The script is small and could be made available if it would be of use to you. It still contains some vestigial iRODS 3.3.1 support, and it embodies some stale local assumptions about where binaries may be found, but these are easy to remove.

So... then I have to munge some pathnames in the json. I would prefer not to have needed to do this, and that is the motivation for this issue.

### Expected behavior

JSON entry `"irods_authentication_file": "/home/ubuntu/.irods/.irodsA-dev",` being written as ``"irods_authentication_file": ".irodsA-dev",`, should cause `iinit`and all consumers of the file to looks in `~/.irods/.irodsA-dev`. Or more generally, a path relative to the JSON file in which the `irods_authentication_file` is named.

### Observed behavior (including steps to reproduce, if applicable)

`iinit` writes `$PWD/.irodsA-dev` relative to the current working directory at time of execution of the command. Tested for 4.1.12 and 4.2.7 by modifying my `irods_environment-$foo.json` and running `iinit`.

## Impact of "fixing" it

I'm afraid I haven't checked the docs to see whether this is existing documented behaviour, or just an undefined behaviour that doesn't suit me. This may affect the likelihood of a "fix" breaking it for other people.

I believe that fixing it would make various types of testing more convenient for developers, so is probably worthwhile.

**Comment by trel** [comment]
Agreed, `iinit`, I would think, should interpret the authentication file relative (if not specified as an absolute path) to the location of the `IRODS_ENVIRONMENT_FILE` file (instead of `iinit`'s $PWD).

I'd lean to categorizing this as a bug at this time.

**Comment by mcast** [comment]
Thanks Trel. Just FYI, this tripped me up for a few minutes but it was good that I had already seen it:

Sysad gave me a config file for a new zone, containing the line `"irods_authentication_file": ".irods/.irods_SSL_A",`

It would work fine if I `cd` home, but I happened to be in `~/bin/` which does not contain a `~/bin/.irods/` and so I got some confusing error messages.

```
ubuntu@mca-icli427:~/bin$ iinit
Enter your current PAM password:
ubuntu@mca-icli427:~/bin$ ils /
Enter your current iRODS password:[-]   /irods/plugins/auth/native/libnative.cpp:342:irods::error native_auth_client_response(irods::plugin_context &, rcComm_t *) :  status [CAT_INVALID_AUTHENTICATION]  errno [] -- message [Call to rcAuthResponseFailed.]

 failed with error -826000 CAT_INVALID_AUTHENTICATION 
```

**Comment by trel** [comment]
In this scenario, where did `iinit` write the authentication file?

And then, I suspect, that is not the location that `ils` went looking... which is why it, too, asked for the (native) password (not PAM).

**Comment by mcast** [comment]
> In this scenario, where did iinit write the authentication file?

I beleive it didn't. Certainly `~/bin/.irods/` was not created, which is fair enough.

Presumably that local `mkdir` failed, and either the failure was not detected or the exception only caused the authentication to fail.

> And then, I suspect, that is not the location that ils went looking... which is why it, too, asked for the (native) password (not PAM).

I assumed `ils` would look in the same place, but can test if you want?

**Comment by trel** [comment]
Yes, anything we can learn to help reproduce and box it in is helpful.   Thanks.

**Comment by mcast** [comment]
Here's a longer transcript:

```
ubuntu@mca-icli427:~$ i-bh9 -v bash
[d] IRODS_ENVIRONMENT_FILE=/home/ubuntu/.irods/irods_environment-bh9.json
[d] exec bash
ubuntu@mca-icli427:~$ cat $IRODS_ENVIRONMENT_FILE 
{
"irods_authentication_file": ".irods/.irods_SSL_A",
"irods_default_hash_scheme": "SHA256",
"irods_authentication_scheme": "PAM",
"irods_cwd": "/bh9_zone/home/mca",
"irods_home": "/bh9_zone/home/mca",
"irods_host": "bh9-terraformed-ies-0.internal.sanger.ac.uk",
"irods_port": 1247,
"irods_ssl_ca_certificate_file": "/etc/ssl/certs/ca-certificates.crt",
"irods_user_name": "mca",
"irods_zone_name": "bh9_zone"
}
ubuntu@mca-icli427:~$ ils /
/:
  C- /bh9_zone
ubuntu@mca-icli427:~$ cd bin
ubuntu@mca-icli427:~/bin$ ils /
Enter your current iRODS password:[-]   /irods/plugins/auth/native/libnative.cpp:342:irods::error native_auth_client_response(irods::plugin_context &, rcComm_t *) :  status [CAT_INVALID_AUTHENTICATION]  errno [] -- message [Call to rcAuthResponseFailed.]

 failed with error -826000 CAT_INVALID_AUTHENTICATION 
ubuntu@mca-icli427:~/bin$ ls -a .irods
ls: cannot access '.irods': No such file or directory
ubuntu@mca-icli427:~/bin$ mkdir .irods
ubuntu@mca-icli427:~/bin$ ils /
Enter your current iRODS password:[-]   /irods/plugins/auth/native/libnative.cpp:342:irods::error native_auth_client_response(irods::plugin_context &, rcComm_t *) :  status [CAT_INVALID_AUTHENTICATION]  errno [] -- message [Call to rcAuthResponseFailed.]

 failed with error -826000 CAT_INVALID_AUTHENTICATION 
ubuntu@mca-icli427:~/bin$ iinit
Enter your current PAM password:
ubuntu@mca-icli427:~/bin$ ils /
/:
  C- /bh9_zone
ubuntu@mca-icli427:~/bin$ rm .irods/.irods_SSL_A 
ubuntu@mca-icli427:~/bin$ ils /
Enter your current iRODS password:[-]   /irods/plugins/auth/native/libnative.cpp:342:irods::error native_auth_client_response(irods::plugin_context &, rcComm_t *) :  status [CAT_INVALID_AUTHENTICATION]  errno [] -- message [Call to rcAuthResponseFailed.]

 failed with error -826000 CAT_INVALID_AUTHENTICATION 
ubuntu@mca-icli427:~/bin$ cd
ubuntu@mca-icli427:~$ ils /
/:
  C- /bh9_zone
ubuntu@mca-icli427:~$ cd bin
ubuntu@mca-icli427:~/bin$ rmdir .irods
ubuntu@mca-icli427:~/bin$ iinit
Enter your current PAM password:
ubuntu@mca-icli427:~/bin$ echo $?
0
ubuntu@mca-icli427:~/bin$ ils /
Enter your current iRODS password:[-]   /irods/plugins/auth/native/libnative.cpp:342:irods::error native_auth_client_response(irods::plugin_context &, rcComm_t *) :  status [CAT_INVALID_AUTHENTICATION]  errno [] -- message [Call to rcAuthResponseFailed.]

 failed with error -826000 CAT_INVALID_AUTHENTICATION 
ubuntu@mca-icli427:~/bin$ ls -l i-bh9 
lrwxrwxrwx 1 ubuntu ubuntu 6 Apr 28 16:43 i-bh9 -> _i-foo
ubuntu@mca-icli427:~/bin$ mkdir .irods
ubuntu@mca-icli427:~/bin$ iinit
Enter your current PAM password:
ubuntu@mca-icli427:~/bin$ diff -u <( hd ~/.irods/.irods_SSL_A ) <( hd ~/bin/.irods/.irods_SSL_A )
--- /dev/fd/63  2020-04-30 19:32:12.551324724 +0000
+++ /dev/fd/62  2020-04-30 19:32:12.551324724 +0000
@@ -1,3 +1,3 @@
-00000000  2e 48 6d [...]
-00000010  4b 22 78 [...]
+00000000  2e 50 6a [...]
+00000010  4e 6c 76 [...]
 0000001b
ubuntu@mca-icli427:~/bin$ ils /
/:
  C- /bh9_zone
ubuntu@mca-icli427:~/bin$ cd
ubuntu@mca-icli427:~$ ils /
/:
  C- /bh9_zone
ubuntu@mca-icli427:~$ rm -r bin/.irods
```

Notes

* `i-bh9` is a prefix command that sets environment, and the reason I was sitting in the `bin` when I first tried to use the zone
* hexdumps elided, but the two authfiles are mostly different from each other
* it seems (at least) two tokens can remain valid concurrently
* when it requests any password I always gave the LDAP password.  I don't have native auth on this zone.
* `iinit` is returning success even when completely failing

**Comment by trel** [comment]
Cursory inspection suggests that `iinit` and `ils` are behaving differently in regard to where they are consulting for credentials.

Can you confirm a different iCommand...  a simple `$ iquest "%s" "select ZONE_NAME"` both behaves and fails in the same way the `ils /` does?
